### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by mahammad-mostafa**
- Several attribute validation rules for `User` & `Post` models
- Attributes (`name` / `title` / `posts_counter` / `comments_counter` / `likes_counter`) are required
- Attribute `title` must be smaller than or equal 250 characters
- Attributes (`posts_counter` / `comments_counter` / `likes_counter`) must be an integer and greate than or equal zero
- Unit tests for attributes validation rules and model methods

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      subject.posts_counter = '-1'
+      expect(subject).to_not(be_valid)
+    end
+  end
+  context 'recent_posts method' do
+    it 'should return last three posts' do
+      expect(subject.recent_posts).to(eq(subject.posts))
+    end
+  end
+end
```
Comment (topeogunleye) [inline spec/models/user_spec.rb]: Good job on your test coverage! Your attention to the required validations in the User model is commendable. You've written tests for the name and posts_counter validations, as well as the recent_posts method, which are crucial for the integrity of the User model.

### Room for Improvement:
However, I noticed that while we have tests for negative values and zero for `posts_counter`, we don't have explicit tests for positive values. It's essential to test the boundary conditions of our validations to make sure they behave correctly across the expected range of inputs.

```.txt
PostsCounter must be an integer greater than or equal to zero.
```
### Suggestion for Improvement:
Consider adding test cases for positive values to ensure comprehensive coverage. For instance, a test for `posts_counter` could be as follows:

```ruby
it 'is valid with a positive likes_counter value' do
    subject.posts_counter = '1'
    expect(post).to be_valid
end
```

Adding such tests will help in confirming that our model not only rejects invalid data but also accepts valid data as it should. This is just as important in TDD (Test-Driven Development) to prevent any false positives and to ensure the robustness of your application.

### Final Note:
Again, great work on your tests so far. By expanding the test cases to cover the full range of valid and invalid inputs, you will further solidify your application's reliability and maintainability.

Keep it up!

---

File: spec/models/post_spec.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  context 'likes_counter attribute' do
+    it 'should be an integer' do
+      subject.likes_counter = 'i'
+      expect(subject).to_not(be_valid)
+    end
+    it 'should be greate than or equal zero' do
+      subject.likes_counter = '-1'
+      expect(subject).to_not(be_valid)
+    end
+  end
```
Comment (topeogunleye) [inline spec/models/post_spec.rb]: Good job on your test coverage! Your attention to the required validations in the Post model is commendable. You've successfully written tests to check for the presence of a title, character limits, and author association, which are crucial for the integrity of the Post model.

### Room for Improvement:
However, I noticed that while we have tests for negative values and zero for `comments_counter` and `likes_counter`, we don't have explicit tests for positive values. It's essential to test the boundary conditions of our validations to make sure they behave correctly across the expected range of inputs.

```.txt
LikesCounter must be an integer greater than or equal to zero.
```
### Suggestion for Improvement:
Consider adding test cases for positive values to ensure comprehensive coverage. For instance, a test for `likes_counter` could be as follows:

```ruby
it 'is valid with a positive likes_counter value' do
       subject.likes_counter = '1'
       expect(post).to be_valid
end
```

Adding such tests will help in confirming that our model not only rejects invalid data but also accepts valid data as it should. This is just as important in TDD (Test-Driven Development) to prevent any false positives and to ensure the robustness of your application.

### Final Note:
Again, great work on your tests so far. By expanding the test cases to cover the full range of valid and invalid inputs, you will further solidify your application's reliability and maintainability.

Keep it up!

---

File: spec/models/post_spec.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
+require 'rails_helper'
+
+RSpec.describe Post, type: :model do
+  user = User.new(name: 'name', photo: 'photo', bio: 'bio', posts_counter: 0)
+  subject { Post.new(author: user, title: 'title', text: 'text', comments_counter: 0, likes_counter: 0) }
```
Comment (topeogunleye) [inline spec/models/post_spec.rb:5]: Great Job, The test you've written to check that a zero `posts_counter` is valid is spot on, and this is critical for preventing invalid data from entering your system ✔️ 💪 💪 💪

**Review by topeogunleye — CHANGES_REQUESTED** [review]
Hi @mahammad-mostafa, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

Try-Again

> ## (Highlights) Good Points: 💪 
### Well-Implemented Aspects:

1. **Negative Value Handling**: You've done an excellent job ensuring that your model validations handle negative numbers. The test you've written to check that a negative `posts_counter` is invalid is spot on, and this is critical for preventing invalid data from entering your system.
2. **Zero Handling**: You've done an excellent job ensuring that your model validations handle zero. The test you've written to check that a zero `posts_counter` is valid is spot on, and this is critical for preventing invalid data from entering your system.

3. **Presence Validation for Name**: You've correctly identified and tested that a user should not be valid without a name. This is a fundamental validation that helps maintain data quality in the application.

4. **Test Clarity**: Your tests are clear and concise, which makes them easy to understand and maintain. This clarity in testing is essential for you and anyone else who might work on this codebase in the future.

5. **Focused Test Cases**: Each of your tests focuses on a single piece of functionality. This focus is a good testing practice because it means that when a test fails, it's easy to pinpoint exactly what went wrong.

6. **Adherence to Requirements**: From the tests provided, it appears you have a good understanding of the requirements and have translated them into specific test cases. This is a great approach to ensure that all the specified requirements are met.


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- For future pull requests, please notice that your branch name `tests` does not follow the convention of being named after a feature or action. For example, you could name your branch `added-tests`. This makes it easier to track the changes and the purpose of each branch.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @topeogunleye in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by ridwanediallo — APPROVED** [review]
# STATUS: APPROVED 🟢
## Hi :wave:  @mahammad-mostafa , 
Congratulations!GIF

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## (Highlights) Good Points: :+1:
- You documented your work professionally. ✔
- You used GitHub flow correctly. ✔
- **Fantastic job, Mahammad!** You have successfully implemented the suggestions. 🎉 🎉 🎉 ✔

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
